src/utils/io/progress_log.rs

```diff
@@ -0,0 +1,235 @@
+        if let Poll::Ready(Ok(buf)) = &result {
+            this.progress.inc(buf.len() as u64);
+        }
+        result
+    }
+
+    fn consume(mut self: Pin<&mut Self>, amt: usize) {
+        Pin::new(&mut self.stream).consume(amt);
+    }
+}
+
+impl<S: futures_core::Stream + Unpin> futures_core::Stream for WithProgressStream<S> {
+    type Item = S::Item;
+
+    fn poll_next(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Option<Self::Item>> {
+        let this = self.get_mut();
+        let item = std::pin::Pin::new(&mut this.stream).poll_next(cx);
+        match &item {
+            std::task::Poll::Ready(Some(_)) => {
+                this.progress.inc(1);
+            }
+            std::task::Poll::Ready(None) => this.progress.finish(),
+            std::task::Poll::Pending => {}
+        }
+        item
+    }
+}
+
+#[derive(Debug, Clone)]
+struct WithProgress {
+    completed_items: u64,
+    frequency: Duration,
+    start: Instant,
+    last_logged: Instant,
+    message: String,
+    total_items: u64,
+}
+
+impl WithProgress {
+    fn new(message: &str, total_items: u64) -> Self {
+        let now = Instant::now();
+        Self {
+            completed_items: 0,
+            frequency: UPDATE_FREQUENCY,
+            start: now,
+            last_logged: now,
+            message: message.into(),
+            total_items,
+        }
+    }
+
+    fn inc(&mut self, value: u64) {
+        self.completed_items += value;
+
+        self.emit_log_if_required();
+    }
+
+    fn set(&mut self, value: u64) {
+        self.completed_items = value;
+
+        self.emit_log_if_required();
+    }
+
+    fn set_total(&mut self, value: u64) {
+        self.total_items = value;
+
+        self.emit_log_if_required();
+    }
+
+    fn emit_log_if_required(&mut self) {
+        let now = Instant::now();
+        if (now - self.last_logged) > self.frequency {
+            let elapsed_secs = (now - self.start).as_secs_f64();
+            let elapsed_duration = format_duration(Duration::from_secs(elapsed_secs as u64));
+
+            let throughput = self.completed_items as f64 / elapsed_secs;
+
+            let eta_secs =
+                (self.total_items.saturating_sub(self.completed_items)) as f64 / throughput;
+            let eta_duration = format_duration(Duration::from_secs(eta_secs as u64));
+
+            info!(
+                "{} {} (elapsed: {}, eta: {})",
+                self.message, self.completed_items, elapsed_duration, eta_duration
+            );
+            self.last_logged = now;
+        }
+    }
+
+    fn finish(&mut self) {}
+}
+
+#[derive(Debug, Clone)]
+pub struct ProgressLog {
+    progress: Arc<Mutex<WithProgress>>,
+}
+
+impl ProgressLog {
+    pub fn new(message: &str, total_items: u64) -> Self {
+        ProgressLog {
+            progress: Arc::new(Mutex::new(WithProgress::new(message, total_items))),
```

Comment:
I would opt for atomic instead of `Mutex`. It's more efficient as atomic ops are supported by the CPU. That way we avoid locking altogether as well as context switching in case we hit a lock.

Reply:
Sure, not blocking.